enforcer-rules/src/main/java/com/google/cloud/tools/opensource/enforcer/LinkageCheckerRule.java

```diff
@@ -0,0 +1,172 @@
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.plugin.logging.Log;
+import org.apache.maven.plugins.enforcer.AbstractNonCacheableEnforcerRule;
+import org.apache.maven.project.DefaultDependencyResolutionRequest;
+import org.apache.maven.project.DependencyResolutionException;
+import org.apache.maven.project.DependencyResolutionRequest;
+import org.apache.maven.project.DependencyResolutionResult;
+import org.apache.maven.project.MavenProject;
+import org.apache.maven.project.ProjectDependenciesResolver;
+import org.codehaus.plexus.component.configurator.expression.ExpressionEvaluationException;
+import org.codehaus.plexus.component.repository.exception.ComponentLookupException;
+import org.eclipse.aether.RepositoryException;
+import org.eclipse.aether.RepositorySystemSession;
+import org.eclipse.aether.artifact.Artifact;
+import org.eclipse.aether.graph.Dependency;
+
+/** Linkage Checker Maven Enforcer Rule. */
+public class LinkageCheckerRule extends AbstractNonCacheableEnforcerRule {
+
+  /**
+   * The section this rule reads dependencies from. By default, it's {@link
+   * DependencySection#DEPENDENCIES}.
+   */
+  private DependencySection dependencySection = DependencySection.DEPENDENCIES;
+
+  @VisibleForTesting
+  void setDependencySection(DependencySection dependencySection) {
+    this.dependencySection = dependencySection;
+  }
+
+  @VisibleForTesting
+  enum DependencySection {
+    /** To read {@code dependencyManagement} section in pom.xml. This is for BOM projects */
+    DEPENDENCY_MANAGEMENT,
+    /** To read {@code dependencies} section in pom.xml. This is for library users' projects */
+    DEPENDENCIES
+  }
+
+  @Override
+  public void execute(@Nonnull EnforcerRuleHelper helper) throws EnforcerRuleException {
+    Log logger = helper.getLog();
+
+    try {
+      MavenProject project = (MavenProject) helper.evaluate("${project}");
+      MavenSession session = (MavenSession) helper.evaluate("${session}");
+      RepositorySystemSession repositorySystemSession = session.getRepositorySession();
+
+      boolean readingDependencyManagementSection =
+          dependencySection == DependencySection.DEPENDENCY_MANAGEMENT;
+      if (readingDependencyManagementSection
+          && (project.getDependencyManagement() == null
+              || project.getDependencyManagement().getDependencies() == null
+              || project.getDependencyManagement().getDependencies().isEmpty())) {
+        logger.warn("The rule is set to read dependency management section but it is empty.");
+      }
+
+      ImmutableList<Path> classpath =
+          readingDependencyManagementSection
+              ? findBomClasspath(project)
+              : findProjectClasspath(project, repositorySystemSession, helper);
+      if (classpath.isEmpty()) {
+        logger.warn("Class path is empty.");
+        return;
+      }
+
+      List<Path> artifactJarsInProject = classpath.subList(0, project.getDependencies().size());
```

Comment:
artifactJarsInProject --> directDependencies
or perhaps add a comment to explain what this is doing. I see it now, but I might not in 6 months.

Reply:
Renamed to directDependencies.